ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java

```diff
@@ -662,17 +690,48 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
     replLogger.tableLog(tblName, tableSpec.tableHandle.getTableType());
     if (tableSpec.tableHandle.getTableType().equals(TableType.EXTERNAL_TABLE)
             || Utils.shouldDumpMetaDataOnly(conf)) {
-      return;
+      return Collections.EMPTY_LIST;
+    }
+    return replPathMappings;
+  }
+
+  private void intitiateDataCopyTasks() {
+    Iterator<ExternalTableCopyTaskBuilder.DirCopyWork> extCopyWorkItr = work.getDirCopyIterator();
+    ReplOperationCompleteAckWork replDumpCompleteAckWork = new ReplOperationCompleteAckWork(dumpAckFile);
+    Task<ReplOperationCompleteAckWork> dumpCompleteAckWorkTask = TaskFactory.get(replDumpCompleteAckWork, conf);
+    List<Task<?>> childTasks = new ArrayList<>();
+    int maxTasks = conf.getIntVar(HiveConf.ConfVars.REPL_APPROX_MAX_LOAD_TASKS);
+    TaskTracker taskTracker = new TaskTracker(maxTasks);
+    while (taskTracker.canAddMoreTasks() && hasMoreCopyWork()) {
+      if (work.replPathIteratorInitialized() && extCopyWorkItr.hasNext()) {
+        childTasks.addAll(new ExternalTableCopyTaskBuilder(work, conf).tasks(taskTracker));
+      } else {
+        childTasks.addAll(ReplPathMapping.tasks(work, taskTracker, conf));
+      }
     }
-    for (ReplPathMapping replPathMapping: replPathMappings) {
-      Task<?> copyTask = ReplCopyTask.getLoadCopyTask(
-              tuple.replicationSpec, replPathMapping.getSrcPath(), replPathMapping.getTargetPath(), conf, false);
-      this.addDependentTask(copyTask);
-      LOG.info("Scheduled a repl copy task from [{}] to [{}]",
-              replPathMapping.getSrcPath(), replPathMapping.getTargetPath());
+    if (!childTasks.isEmpty()) {
+      boolean ackTaskAdded = false;
+      if (taskTracker.canAddMoreTasks()) {
+        childTasks.add(dumpCompleteAckWorkTask);
+        ackTaskAdded = true;
+      }
+      if (hasMoreCopyWork() || !ackTaskAdded) {
+        DAGTraversal.traverse(childTasks, new AddDependencyToLeaves(TaskFactory.get(work, conf)));
```

Comment:
Add the ack only after no more work is pending

Reply:
Ack is being added at the end in these cases as well